### DISCUSSION TITLE
SQL Alchemy Schema Design

### DISCUSSION THREAD

**Discussion by rohinb2** [sqlalchemy/sqlalchemy#6774] [Usage Questions]
Somewhat of a beginner question - I understand how to solve my problem using general SQL but had trouble piecing the docs together to figure out the right avenue for solving this.

I have a rather complex relationship I need to model and wanted to see if anyone had any thoughts on it. Currently, I have a SQL alchemy DB with a schema along the lines of:

```
Main:
    id

A:
    id
    name
    (fk) main.id
    (unique constraint) name, main.id

B:
    id
    name
    (fk) main.id
    (unique constraint) name, main.id

# 3 more tables of the same format as A and B
```

This models a one-to-many relationship from Main to A, Main to B, etc

However, I'd like these to be changed to many-to-many relationships, while also enforcing the unique constraint by name. I was trying to see if this could be accomplished with one complex junction table, rather than a separate junction table for Main to A, Main to B, etc

Something along the lines of:

```
Main:
    id

MainJunction:
    main.id
    table
    table.id
    table.name
     (unique constraint) main.id, table, table.id, table.name

A:
    id
    name

B:
    id
    name

# 3 more tables of the same format as A and B
```
I was having trouble trying to understand the best way to model this in SQL alchemy such that I don't have to write custom joins for everything and I can still utilize the features of the ORM itself (relationships, backref, backpopulates, etc).

Let me know if there's any other info I can provide. I read most of [this page](https://docs.sqlalchemy.org/en/14/orm/join_conditions.html) but couldn't figure out how to model this relationship properly.

**Comment by CaselIT** [comment]
Hi,

To use a many to many you usually have to specify the secondary parameter of the relationship. See here https://docs.sqlalchemy.org/en/14/orm/relationship_api.html#sqlalchemy.orm.relationship.params.secondary for the docs. there are also a many see-other links that may be useful.

Note that I think you will need to manually specify the secondary join, since in `MainJunction` `table.id` cannot be a fk to the child tables, since it needs to target more than one table.
Personally I think it would be easier to just specify multiple secondary table, so you can retain the db foreign key and proper cascade, but that's only my opinion.

**Reply by rohinb2** [reply]
Thanks for the help. Is there any documentation on the easiest way to migrate a one-to-many relationship to a many-to-many relationship?
